webapp/application.py

```diff
@@ -125,4 +131,114 @@ def application_page(token):
+
+    if not (email and candidate_id and application_id):
+        flask.abort(404)
+
+    application = harvest.get_application(application_id)
+    candidate = harvest.get_candidate(application["candidate_id"])
+    if (
+        "candidate_id" not in application
+        or application["candidate_id"] != candidate_id
+    ):
+        flask.abort(404)
+
+    print(candidate["first_name"], withdrawal_reason)
+
+    # TODO: call the Greenhouse API to reject the application here
+    # ...
+    return flask.render_template("applications/withdrawal.html")
+
+
+@application.route("/withdraw/<string:token>", methods=["POST"])
+def sendForm(token):
+    decrypted = cipher.decrypt(token)
+    if not decrypted:
+        flask.abort(404)
+    name, candidate_id, application_id = tuple(decrypted.split("-"))
+    candidate_id, application_id = int(candidate_id), int(application_id)
+    application = harvest.get_application(application_id)
+    candidate = harvest.get_candidate(application["candidate_id"])
+    if (
+        "candidate_id" not in application
+        or application["candidate_id"] != candidate_id
+        or not candidate["first_name"].lower() == name
+    ):
+        flask.abort(404)
+
+    candidate_name = candidate["first_name"]
+    position = application["jobs"][0]["name"]
+
+    # Sanitize and parse user input
+    email = parseaddr(flask.request.form["email"])[1]
+    candidate_email = parseaddr(candidate["email_addresses"][0]["value"])[1]
+    textarea = html.escape(flask.request.form["textarea"])
+
+    # Reject if user typed the wrong email
+    if not candidate_email == email:
+        flask.abort(404)
+
+    send_mail(
+        to_email=[candidate_email],
+        subject="Withdraw Application Confirmation",
+        message=flask.render_template(
+            "applications/_activate-email.html",
+            applicant_name=candidate_name,
+            position=position,
+            verification_link=confirmation_token(
+                candidate_email, textarea, candidate_id, application_id
+            ),
+        ),
     )
+    return redirect(flask.request.referrer + "#withdrawal-requested")
+
+
+def confirmation_token(email, withdrawal_reason, candidate_id, application_id):
+    payload = {
+        "email": email,
+        "withdrawal_reason": withdrawal_reason,
+        "candidate_id": candidate_id,
+        "application_id": application_id,
+    }
+    token = json.dumps(payload)
+    return verification_token_cipher.encrypt(token)
+
+
+def send_mail(
+    to_email,
+    subject,
+    message,
+    server="smtp.gmail.com",
+    from_email="min.kim@canonical.com",
+):
+    # import smtplib
+    try:
+        msg = EmailMessage()
+        msg["Subject"] = subject
+        msg["From"] = from_email
+        msg["To"] = ", ".join(to_email)
+        msg.set_content(message)
+
+        server = smtplib.SMTP(server, 587)
+        server.starttls()
+        server.login(from_email, "")  # user & password
```

Comment:
Isn't a password required? How is it going to get a password? Or does smtp.gmail.com literally let you send without one?

Presumably the from email isn't going to be min.kim. What's the ultimate email going to be? And will it actually work with the gmail SMTP server?